docker/docker.ts

```diff
@@ -397,48 +397,57 @@ async function dockerBuild(
     }
 
     await runCommandThatMustSucceed("docker", buildArgs, logResource);
+
+}
+
+interface LoginResult {
+    registryName: string;
+    username: string;
+    loginCommand: Promise<void> | undefined;
 }
 
 // Keep track of registries and users that have been logged in.  If we've already logged into that
 // registry with that user, there's no need to do it again.
-const loggedInUsers: { registryName: string, username: string }[] = [];
+const loginResults: LoginResult[] = [];
 
-async function loginToRegistry(registry: Registry, logResource: pulumi.Resource): Promise<void> {
+function loginToRegistry(registry: Registry, logResource: pulumi.Resource): Promise<void> {
     const { registry: registryName, username, password } = registry;
 
-    if (isLoggedIn(registry)) {
-        logEphemeral(`Reusing existing login for ${username}@${registryName}`, logResource);
-        return;
-    }
-
-    loggedInUsers.push({ registryName, username });
-
-    const dockerPasswordStdin = await useDockerPasswordStdin(logResource);
-
-    // pass 'reportFullCommandLine: false' here so that if we fail to login we don't emit the
-    // username/password in our logs.  Instead, we'll just say "'docker login' failed with code ..."
-    if (dockerPasswordStdin) {
-        await runCommandThatMustSucceed(
-            "docker", ["login", "-u", username, "--password-stdin", registryName],
-            logResource, /*reportFullCommandLine*/ false, password);
+    // See if we've issued an outstanding requests to login into this registry.  If so, just
+    // await the results of that login request.  Otherwise, create a new request and keep it
+    // around so that future login requests will see it.
+    let loginResult = loginResults.find(
+        r => r.registryName === registryName && r.username === username);
+    if (!loginResult) {
+        // Note: we explicitly do not 'await' the 'loginAsync' call here.  We do not want
+        // to relinquish control of this thread-of-execution yet.  We want to ensure that
+        // we first update `loginResults` with our record object so that any future executions
+        // through this method see that the login was kicked off and can wait on that.
+        loginResult = { registryName, username, loginCommand: loginAsync() };
+        loginResults.push(loginResult);
     }
     else {
-        await runCommandThatMustSucceed(
-            "docker", ["login", "-u", username, "-p", password, registryName],
-            logResource, /*reportFullCommandLine*/ false);
+        logEphemeral(`Reusing existing login for ${username}@${registryName}`, logResource);
     }
-}
 
-function isLoggedIn(registry: Registry): boolean {
-    for (const { registryName, username } of loggedInUsers) {
-        if (registryName === registry.registry &&
-            username === registry.username) {
+    return loginResult.loginCommand!;
```

Comment:
Why is the `!` here safe?  Is the `|undefined` actually necessary in the type?

Reply:
you're right.  will fix.